### PR TITLE
Translate Backend DL Clients 

### PR DESCRIPTION
**Pull request by stevietv**
#### Database Migration
NO

#### Description
Adds translations to most download client UI fields served by the backend.
Exceptions are not translated
validation rules present in the settings.cs files are not translated

#### Todos

Lemme know if you want it squashed!

#### Issues Fixed or Closed by this PR

*

### CODE REVIEW

**Comment by stevietv** [comment]
So whilst this works for all the UI work I've done, it seems that these three lines cause a whole bunch of Integration and API tests to fail with an `System.NullReferenceException : Object reference not set to an instance of an object.`

https://github.com/Sonarr/Sonarr/blob/7f58bea7aee314d5b99438851eb0d5280cbd5925/src/Sonarr.Http/ClientSchema/SchemaBuilder.cs#L121-L124

Leaving these three lines as:
```
    Label = fieldAttribute.Label,
    Unit = fieldAttribute.Unit,
    HelpText = fieldAttribute.HelpText,
    HelpTextWarning = fieldAttribute.HelpTextWarning
```

Means that all the tests pass again, but then of course the keys in the attributes don't get translated in the UI passed.
Not sure of the best way forward here?

**Comment by stevietv** [comment]
figured it out. `_localizationService` wasn't initialized in the context of the tests, so added a check for not null in the places where its used.
Tests all pass now (including a tweaked Hadouken test)/

**Review by markus101 — CHANGES_REQUESTED** [review]
As mentioned in the comment we should fix the tests instead of bypassing the localization service. These diffs will fix the tests that I saw were broken originally.

```
diff --git a/src/NzbDrone.Api.Test/ClientSchemaTests/SchemaBuilderFixture.cs b/src/NzbDrone.Api.Test/ClientSchemaTests/SchemaBuilderFixture.cs
index d3c036327..6af8abb7b 100644
--- a/src/NzbDrone.Api.Test/ClientSchemaTests/SchemaBuilderFixture.cs
+++ b/src/NzbDrone.Api.Test/ClientSchemaTests/SchemaBuilderFixture.cs
@@ -1,6 +1,9 @@
+using System.Collections.Generic;
 using FluentAssertions;
+using Moq;
 using NUnit.Framework;
 using NzbDrone.Core.Annotations;
+using NzbDrone.Core.Localization;
 using NzbDrone.Test.Common;
 using Sonarr.Http.ClientSchema;
 
@@ -9,6 +12,16 @@ namespace NzbDrone.Api.Test.ClientSchemaTests
     [TestFixture]
     public class SchemaBuilderFixture : TestBase
     {
+        [SetUp]
+        public void Setup()
+        {
+            Mocker.GetMock<ILocalizationService>()
+                .Setup(s => s.GetLocalizedString(It.IsAny<string>(), It.IsAny<Dictionary<string, object>>()))
+                .Returns<string, Dictionary<string, object>>((s, d) => s);
+
+            SchemaBuilder.Initialize(Mocker.Container);
+        }
+
         [Test]
         public void should_return_field_for_every_property()
         {
```

```
diff --git a/src/NzbDrone.Integration.Test/IntegrationTest.cs b/src/NzbDrone.Integration.Test/IntegrationTest.cs
index a6721fdcb..9524d39d3 100644
--- a/src/NzbDrone.Integration.Test/IntegrationTest.cs
+++ b/src/NzbDrone.Integration.Test/IntegrationTest.cs
@@ -1,3 +1,5 @@
+using System;
+using System.Linq;
 using System.Threading;
 using NLog;
 using NUnit.Framework;
@@ -7,7 +9,6 @@
 using NzbDrone.Core.Indexers.Newznab;
 using NzbDrone.Test.Common;
 using NzbDrone.Test.Common.Datastore;
-using Sonarr.Http.ClientSchema;
 
 namespace NzbDrone.Integration.Test
 {
@@ -50,17 +51,20 @@ protected override void InitializeTestTarget()
             // Make sure tasks have been initialized so the config put below doesn't cause errors
             WaitForCompletion(() => Tasks.All().SelectList(x => x.TaskName).Contains("RssSync"));
 
-            Indexers.Post(new Sonarr.Api.V3.Indexers.IndexerResource
+            var indexer = Indexers.Schema().FirstOrDefault(i => i.Implementation == nameof(Newznab));
+
+            if (indexer == null)
             {
-                EnableRss = false,
-                EnableInteractiveSearch = false,
-                EnableAutomaticSearch = false,
-                ConfigContract = nameof(NewznabSettings),
-                Implementation = nameof(Newznab),
-                Name = "NewznabTest",
-                Protocol = Core.Indexers.DownloadProtocol.Usenet,
-                Fields = SchemaBuilder.ToSchema(new NewznabSettings())
-            });
+                throw new NullReferenceException("Expected valid indexer schema, found null");
+            }
+
+            indexer.EnableRss = false;
+            indexer.EnableInteractiveSearch = false;
+            indexer.EnableAutomaticSearch = false;
+            indexer.ConfigContract = nameof(NewznabSettings);
+            indexer.Implementation = nameof(Newznab);
+            indexer.Name = "NewznabTest";
+            indexer.Protocol = Core.Indexers.DownloadProtocol.Usenet;
 
             // Change Console Log Level to Debug so we get more details.
             var config = HostConfig.Get(1);
```

```
diff --git a/src/NzbDrone.Test.Common/AutoMoq/AutoMoqer.cs b/src/NzbDrone.Test.Common/AutoMoq/AutoMoqer.cs
index 6e6d84666..948994c51 100644
--- a/src/NzbDrone.Test.Common/AutoMoq/AutoMoqer.cs
+++ b/src/NzbDrone.Test.Common/AutoMoq/AutoMoqer.cs
@@ -23,6 +23,8 @@ public AutoMoqer()
             SetupAutoMoqer(CreateTestContainer(new Container()));
         }
 
+        public IContainer Container => _container;
+
         public virtual T Resolve<T>()
         {
             var result = _container.Resolve<T>();
```

File: src/Sonarr.Http/ClientSchema/SchemaBuilder.cs [hunk 1]
```diff
@@ -110,10 +118,10 @@ private static FieldMapping[] GetFieldMapping(Type type, string prefix, Func<obj
                     var field = new Field
                     {
                         Name = prefix + GetCamelCaseName(propertyInfo.Name),
-                        Label = fieldAttribute.Label,
+                        Label = (fieldAttribute.Label.IsNotNullOrWhiteSpace() && _localizationService != null) ? _localizationService.GetLocalizedString(fieldAttribute.Label, GetTokens(type, fieldAttribute.Label, TokenField.Label)) : fieldAttribute.Label,
                         Unit = fieldAttribute.Unit,
-                        HelpText = fieldAttribute.HelpText,
-                        HelpTextWarning = fieldAttribute.HelpTextWarning,
+                        HelpText = (fieldAttribute.HelpText.IsNotNullOrWhiteSpace()  && _localizationService != null) ? _localizationService.GetLocalizedString(fieldAttribute.HelpText, GetTokens(type, fieldAttribute.Label, TokenField.HelpText)) : fieldAttribute.HelpText,
```
[the remark this answers is not in the document]
Comment (stevietv) [reply src/Sonarr.Http/ClientSchema/SchemaBuilder.cs]: this looks a lot more readable, thanks for the help with getting the tests working, that was my first approach but couldn't figure it out

**Comment by stevietv** [comment]
all changes made and tests pass.

**Review by markus101 — APPROVED** [review]

**Pull request merged**